sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/transform/BeamJoinTransforms.java

```diff
@@ -44,26 +46,48 @@
 
   /** A {@code SimpleFunction} to extract join fields from the specified row. */
   public static class ExtractJoinFields extends SimpleFunction<Row, KV<Row, Row>> {
-    private final List<Integer> joinColumns;
+    private final List<SerializableRexNode> joinColumns;
     private final Schema schema;
+    private int leftRowColumnCount;
 
     public ExtractJoinFields(
-        boolean isLeft, List<Pair<Integer, Integer>> joinColumns, Schema schema) {
+        boolean isLeft,
+        List<Pair<RexNode, RexNode>> joinColumns,
+        Schema schema,
+        int leftRowColumnCount) {
       this.joinColumns =
-          joinColumns.stream().map(pair -> isLeft ? pair.left : pair.right).collect(toList());
+          joinColumns.stream()
+              .map(pair -> SerializableRexNode.builder(isLeft ? pair.left : pair.right).build())
+              .collect(toList());
       this.schema = schema;
+      this.leftRowColumnCount = leftRowColumnCount;
     }
 
     @Override
     public KV<Row, Row> apply(Row input) {
-      Row row = joinColumns.stream().map(input::getValue).collect(toRow(schema));
+      Row row =
+          joinColumns.stream()
+              .map(v -> getValue(v, input, leftRowColumnCount))
+              .collect(toRow(schema));
       return KV.of(row, input);
     }
 
+    @SuppressWarnings("unused")
     private Schema.Field toField(Schema schema, Integer fieldIndex) {
       Schema.Field original = schema.getField(fieldIndex);
       return original.withName("c" + fieldIndex);
     }
+
+    private Object getValue(
```

Comment:
This supports only one level nesting. How hard would it be to implement the linked list of the field indices? E.g. `serializableRexNode.isNested()`, `serializableRexNode.getNestedSerializableRexNode()`. This way the logic can probably be generalized a bit. And you probably won't need the class hierarchy, it would be a linked list with an int payload, and you stop when you reach the end of the list.

Reply:
That's a good idea. 
InputRef -> Index
FieldAccess -> List of Index

Changed.